### PR TITLE
Reshape-able SliceConverter

### PR DESCRIPTION
**Pull request by yekruglov**
Description: Reshape-able convert Slice to StridedSlice

JIRA: 41517

Code:
* [x]  Comments
* [x]  Code style (PEP8)
* [x]  Transformation generates reshape-able IR
* [x]  Transformation preserves original framework node names

Validation:
* [x]  Unit tests
* [x]  Framework operation tests N/A: no new operations were enabled
* [x]  Transformation tests
* [x]  e2e model test with an update of ./tests/e2e_oss/utils/reshape_utils.py N/A: no new models were enabled
* [x]  Model Optimizer IR Reader check

Documentation:
* [x]  Supported frameworks operations list N/A: no new operations were enabled
* [x]  Supported **public** models list N/A: no new models were enabled
* [x]  New operations specification N/A: no new operations were enabled
* [x]  Guide on how to convert the **public** model N/A: no new models were enabled
* [x]  User guide update N/A

### CODE REVIEW

**Comment by lazarevevgeny** [comment]
@yekruglov , update the PR description according to template.

File: model-optimizer/extensions/middle/SliceConverter.py [hunk 1]
```diff
@@ -36,80 +63,59 @@ class ConvertSlice(MiddleReplacementPattern):
     """
 
     enabled = True
-    op = "Slice"
     force_clean_up = True
+    op = "Slice"
```
Comment (jane-intel) [inline model-optimizer/extensions/middle/SliceConverter.py:67]: useless

File: model-optimizer/extensions/middle/SliceConverter.py [hunk 2]
```diff
@@ -16,18 +16,45 @@
...
 
-def convert_negative_indices(indices: np.array, shape: np.array):
-    for ind, value in enumerate(indices):
-        if value < 0:
-            indices[ind] += shape[ind]
+def create_ss_interval_border(graph: Graph, slice_border_port: Port, shape, axes, node_name):
+    # the value for 'ends' might be maximum/minimum possible value of int64. This
+    # value must be converted to maximum/minimum of int32 because such big values do not fit into the int32 which is
+    # supported by the StridedSlice layer
+    clamp = create_op_with_const_inputs(graph, Clamp, port_value_dict={1: np.iinfo(np.int32).min,
```
Comment (jane-intel) [inline model-optimizer/extensions/middle/SliceConverter.py:37]: we clamp for both borders, please correct the comment

File: model-optimizer/extensions/middle/SliceConverter.py [hunk 3]
```diff
@@ -16,18 +16,45 @@
...
 from mo.ops.const import Const
 from mo.ops.strided_slice import StridedSlice
-from mo.utils.error import Error
 
 
-def convert_negative_indices(indices: np.array, shape: np.array):
-    for ind, value in enumerate(indices):
-        if value < 0:
-            indices[ind] += shape[ind]
+def create_ss_interval_border(graph: Graph, slice_border_port: Port, shape, axes, node_name):
```
Comment (jane-intel) [inline model-optimizer/extensions/middle/SliceConverter.py:33]: please add type anotations

File: model-optimizer/extensions/middle/SliceConverter.py [hunk 4]
```diff
@@ -16,18 +16,45 @@
...
+    # the value for 'ends' might be maximum/minimum possible value of int64. This
+    # value must be converted to maximum/minimum of int32 because such big values do not fit into the int32 which is
+    # supported by the StridedSlice layer
+    clamp = create_op_with_const_inputs(graph, Clamp, port_value_dict={1: np.iinfo(np.int32).min,
+                                                                       2: np.iinfo(np.int32).max},
+                                        op_attrs=dict(name=node_name + '/Clamp'))
```
Comment (jane-intel) [inline model-optimizer/extensions/middle/SliceConverter.py:39]: nit: looks more readable to me. here and later
```suggestion
    clamp = create_op_with_const_inputs(
        graph, Clamp, port_value_dict={1: np.iinfo(np.int32).min, 2: np.iinfo(np.int32).max}, op_attrs=dict(name=node_name + '/Clamp'))
```

File: model-optimizer/extensions/middle/SliceConverter.py [hunk 5]
```diff
@@ -16,18 +16,45 @@
...
+    clamp.in_port(0).connect(slice_border_port)
+    cast = Cast(graph, dict(name=node_name + '/CastToI64', dst_type=np.int64)).create_node()
+    cast.in_port(0).connect(clamp.out_port(0))
+    concat = Concat(graph, dict(name=node_name + '/Concat', axis=0)).create_node()
+    for value_idx, port_idx in enumerate(axes):
+        concat.add_input_port(port_idx)
+        value = create_op_with_const_inputs(graph, Gather, port_value_dict={1: int64_array([value_idx]),
+                                                                            2: 0},
```
Comment (jane-intel) [inline model-optimizer/extensions/middle/SliceConverter.py:47]: nit:
```suggestion
                                                                            2: int64_array(0)},
```

File: model-optimizer/extensions/middle/SliceConverter.py [hunk 6]
```diff
@@ -16,18 +16,45 @@
...
+                                            op_attrs={'name': node_name + '/Gather'})
+        cast.out_port(0).connect(value.in_port(0))
+        value.out_port(0).connect(concat.in_port(port_idx))
+    for port_idx in range(len(shape)):
+        if not concat.is_in_port_connected(port_idx):
+            concat.add_input_port(port_idx)
+            const = Const(graph, dict(name=node_name + '/Const', value=int64_array([0]))).create_node()
```
Comment (jane-intel) [inline model-optimizer/extensions/middle/SliceConverter.py:54]: Please add a comment that this border value would be ignored using mask

File: model-optimizer/extensions/middle/SliceConverter.py [hunk 7]
```diff
@@ -36,80 +63,59 @@ class ConvertSlice(MiddleReplacementPattern):
     """
```
Comment (jane-intel) [inline model-optimizer/extensions/middle/SliceConverter.py:63]: Not informative

File: model-optimizer/extensions/middle/SliceConverter.py [hunk 8]
```diff
@@ -36,80 +63,59 @@ class ConvertSlice(MiddleReplacementPattern):
...
 
     enabled = True
-    op = "Slice"
     force_clean_up = True
+    op = "Slice"
 
     def run_after(self):
         from extensions.middle.pass_separator import MiddleStart
         return [MiddleStart]
```
Comment (jane-intel) [inline model-optimizer/extensions/middle/SliceConverter.py:71]: Please delete, as this code is useless -- `MiddleReplacementPattern` already has exactly the same method

File: model-optimizer/extensions/middle/SliceConverter.py [hunk 9]
```diff
@@ -16,100 +16,103 @@
...
-            indices[ind] += shape[ind]
+def create_ss_interval_border(graph: Graph, slice_border_port: Port, shape: np.ndarray, axes: np.ndarray, node_name: str):
+    # the value for 'starts' or 'ends' might be maximum/minimum possible value of int64. This
+    # value must be converted to maximum/minimum of int32 because such big values do not fit into the int32 which is
+    # supported by the StridedSlice layer
+    clamp = create_op_with_const_inputs(
+        graph, Clamp, port_value_dict={1: np.iinfo(np.int32).min, 2: np.iinfo(np.int32).max},
+        op_attrs=dict(name=node_name + '/Clamp'))
+    clamp.in_port(0).connect(slice_border_port)
+    cast = Cast(graph, dict(name=node_name + '/CastToI64', dst_type=np.int64)).create_node()
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/SliceConverter.py:41]: Why do we need this `Cast`? `Gather` supports int32 indices as well.

File: model-optimizer/extensions/middle/SliceConverter.py [hunk 10]
```diff
@@ -16,100 +16,103 @@
...
 
 
-def convert_negative_indices(indices: np.array, shape: np.array):
-    for ind, value in enumerate(indices):
-        if value < 0:
-            indices[ind] += shape[ind]
+def create_ss_interval_border(graph: Graph, slice_border_port: Port, shape: np.ndarray, axes: np.ndarray, node_name: str):
+    # the value for 'starts' or 'ends' might be maximum/minimum possible value of int64. This
+    # value must be converted to maximum/minimum of int32 because such big values do not fit into the int32 which is
+    # supported by the StridedSlice layer
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/SliceConverter.py:36]: Also explain in the comment why you add `Gather` in the code below.

File: model-optimizer/extensions/middle/SliceConverter.py (hunk 9 above)
Comment (lazarevevgeny) [reply model-optimizer/extensions/middle/SliceConverter.py:41]: After reading the code below I understood why this `Cast` is needed. Please, add a description in the function why it is needed.

**Review by lazarevevgeny — COMMENTED** [review]
In general looks good. But need to add some comments describing what happens in the transformation and why.

**Review by jane-intel — COMMENTED** [review]
Comments are the same as @lazarevevgeny just left
In general looks good

**Comment by yekruglov** [comment]
Do we need more comments in this transformation?

**Review by lazarevevgeny — APPROVED** [review]

**Comment by lazarevevgeny** [comment]
> Comments are the same as @lazarevevgeny just left
> In general looks good

Looks good to me. One more thing. Please, re-run layer tests to make sure that none of them are broken.

**Pull request merged**